config.go

```diff
@@ -0,0 +1,329 @@
+// handle all struct tags supported by gopkg.in/yaml.v2. See
+// https://godoc.org/gopkg.in/yaml.v2#Marshal for details.
+type YAML struct {
+	name     string
+	raw      []byte
+	contents interface{}
+	strict   bool
+}
+
+// NewYAML constructs a YAML provider. See the various YAMLOptions for
+// available tweaks to the default behavior.
+func NewYAML(options ...YAMLOption) (y *YAML, retErr error) {
+	cfg := &config{
+		strict: true,
+		name:   "YAML",
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, cfg.close())
+	}()
+	for _, o := range options {
+		o.apply(cfg)
+	}
+
+	if cfg.err != nil {
+		return nil, fmt.Errorf("error applying options: %v", cfg.err)
+	}
+
+	// On construction, go through a full merge-serialize-deserialize cycle to
+	// catch any duplicated keys as early as possible (in strict mode). It also
+	// strips comments, which stops us from attempting environment variable
+	// expansion.
+	merged, err := merge.YAML(cfg.sources, cfg.strict)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't merge YAML sources: %v", err)
+	}
+
+	if cfg.lookup != nil {
+		exp, err := ioutil.ReadAll(transform.NewReader(merged, newExpandTransformer(cfg.lookup)))
+		if err != nil {
+			return nil, fmt.Errorf("couldn't expand environment: %v", err)
+		}
+		merged = bytes.NewBuffer(exp)
+	}
+
+	raw := merged.Bytes() // helpful for WithDefault later on
+	var contents interface{}
+	dec := yaml.NewDecoder(merged)
+	dec.SetStrict(cfg.strict)
+	if err := dec.Decode(&contents); err != nil {
+		return nil, fmt.Errorf("couldn't decode merged YAML: %v", err)
+	}
+
+	return &YAML{
+		name:     cfg.name,
+		raw:      raw,
+		contents: contents,
+		strict:   cfg.strict,
+	}, nil
+}
+
+// Name returns the name of the provider. It defaults to "YAML".
+func (y *YAML) Name() string {
+	return y.name
+}
+
+// Get retrieves a value from the configuration. The supplied key is treated
+// as a period-separated path, with each path segment used as a map key. For
+// example, if the provider contains the YAML
+//   foo:
+//     bar:
+//       baz: hello
+// then Get("foo.bar") returns a value holding
+//   baz: hello
+//
+// To get a value holding the entire configuration, use the Root constant as
+// the key.
+func (y *YAML) Get(key string) Value {
+	return y.get(strings.Split(key, _separator))
+}
+
+func (y *YAML) get(path []string) Value {
+	if len(path) == 1 && path[0] == "" {
+		path = nil // root node
+	}
+	return Value{
+		path:     path,
+		provider: y,
+	}
+}
+
+// at returns the unmarshalled representation of the value at a given path,
+// with a bool indicating whether the value was found.
+func (y *YAML) at(path []string) (interface{}, bool) {
+	cur := y.contents
+	for _, segment := range path {
+		m, ok := cur.(map[interface{}]interface{})
+		if !ok {
+			return nil, false
+		}
+		if _, ok := m[segment]; !ok {
+			return nil, false
+		}
+		cur = m[segment]
+	}
+	return cur, true
+}
+
+func (y *YAML) populate(path []string, i interface{}) error {
```

Comment:
For strict mode: Do we check if all the keys in `interface{}` are populated here?

Reply:
No, that's built into the YAML library. When we call `SetStrict` on decoders, we enable these checks.